### PR TITLE
Convert nightly upload script to Python

### LINKED ISSUE #16560: Windows nightly uploads are failing
http://build.servo.org/builders/windows-msvc-nightly/builds/120/steps/shell__2/logs/stdio

### PR DESCRIPTION
**Pull request by aneeshusa**
Now that MinGW and MSYS have been removed from the Windows builders,
bash is not available to run the previous upload_nightlies.sh script.
Convert the script to Python 2 for cross-platform support.
Additionally, switch to the `boto3` library for easy uploading
without needing to install `s3cmd`,
and move the code into mach for easy `boto3` installation
as the new `./mach upload-nightly` command.

Also, hard-code the paths to the packages instead of using
globs to look for them, as the paths are static.
(The paths used to contain timestamps,
but we now insert timestamps when uploading to S3
to improve reproducibility.)

---

- [ ] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #16560 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because Buildbot will test them.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16565)

### CODE REVIEW

File: etc/ci/upload_nightly [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    subprocess.check_call([
+        'echo',
+        's3cmd',
+        '--mime-type=application/octet-stream',
+        'put',
+        package,
+        package_upload_path,
+    ])
+    subprocess.check_call([
+        'echo',
```
Comment (metajack) [inline etc/ci/upload_nightly]: Are these calls supposed to run s3cmd not echo?

File: etc/ci/upload_nightly [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    )
+    with open(package) as p:
+        digest = hashlib.sha256(p.read()).hexdigest()
+
+    brew_version = timestamp.strftime('%Y.%m.%d')
+
+    with tempfile.TemporaryDirectory(prefix='homebrew-servo') as tmp_dir:
+        def call_git(cmd, **kwargs):
+            subprocess.check_call(
+                ['echo', 'git', '-C', tmp_dir] + cmd,
```
Comment (metajack) [inline etc/ci/upload_nightly]: Same here.

**Review by metajack — CHANGES_REQUESTED** [review]

**Comment by aneeshusa** [comment]
s3cmd apparently is not (well) supported on Windows, so I'll rework this to use a different tool.

**Comment by metajack** [comment]
Can we just use the aws cli tool directly?

**Comment by metajack** [comment]
@aneeshusa I'm happy to drive this across the line if you are busy, btw.

**Comment by aneeshusa** [comment]
@metajack I would appreciate that! Here's where I was thinking of taking this:

- Avoid installing anything globally/pushing anything via saltfs.
- Port this to Python 2 and move this into `mach`.
- Use `boto3` as a python library instead of `s3cmd` via the command line. Install `boto3` as part of the `mach` Python requirements.

This should hopefully sidestep any Windows compatibility issues.

**Comment by aneeshusa** [comment]
I've done the steps from my last comment and this is ready for review. I did some light smoke testing, but haven't tested actually uploading to S3 or the `update_brew` code.

**Comment by atouchet** [comment]
Is this still a WIP or can that tag be removed?

**Comment by larsbergstrom** [comment]
@metajack @edunham 

It looks to me like this is good to go. Should we r+ and then check to see what happens on the nightly builds?

cc @jonathandturner

**Comment by metajack** [comment]
Great work @aneeshusa. This looks like a huge improvement.

**Pull request merged**